domain/src/test/java/org/oppia/android/domain/exploration/ExplorationActiveTimeControllerTest.kt

```diff
@@ -0,0 +1,219 @@
+import org.oppia.android.domain.oppialogger.analytics.ApplicationLifecycleModule
+import org.oppia.android.domain.topic.TEST_TOPIC_ID_0
+import org.oppia.android.testing.FakeExceptionLogger
+import org.oppia.android.testing.TestLogReportingModule
+import org.oppia.android.testing.data.DataProviderTestMonitor
+import org.oppia.android.testing.robolectric.RobolectricModule
+import org.oppia.android.testing.threading.TestCoroutineDispatchers
+import org.oppia.android.testing.threading.TestDispatcherModule
+import org.oppia.android.testing.time.FakeOppiaClock
+import org.oppia.android.testing.time.FakeOppiaClockModule
+import org.oppia.android.util.data.DataProvidersInjector
+import org.oppia.android.util.data.DataProvidersInjectorProvider
+import org.oppia.android.util.locale.testing.LocaleTestModule
+import org.oppia.android.util.logging.EnableConsoleLog
+import org.oppia.android.util.logging.EnableFileLog
+import org.oppia.android.util.logging.GlobalLogLevel
+import org.oppia.android.util.logging.LogLevel
+import org.robolectric.annotation.Config
+import org.robolectric.annotation.LooperMode
+import java.util.concurrent.TimeUnit
+import javax.inject.Inject
+import javax.inject.Singleton
+
+private const val CACHE_NAME = "topic_learning_time_database"
+private const val SESSION_LENGTH = 5000L
+
+/** Tests for [ExplorationActiveTimeController]. */
+@RunWith(AndroidJUnit4::class)
+@LooperMode(LooperMode.Mode.PAUSED)
+@Config(application = ExplorationActiveTimeControllerTest.TestApplication::class)
+class ExplorationActiveTimeControllerTest {
+  @Inject lateinit var fakeExceptionLogger: FakeExceptionLogger
+  @Inject lateinit var testCoroutineDispatchers: TestCoroutineDispatchers
+  @Inject lateinit var monitorFactory: DataProviderTestMonitor.Factory
+  @Inject lateinit var cacheFactory: PersistentCacheStore.Factory
+  @Inject lateinit var oppiaClock: FakeOppiaClock
+  @Inject lateinit var explorationActiveTimeController: ExplorationActiveTimeController
+
+  private val profileId = ProfileId.newBuilder().setInternalId(0).build()
+
+  @Before
+  fun setUp() {
+    setUpTestApplicationComponent()
+  }
+
+  @Test
+  fun testExplorationSessionStarted_setsStartExplorationTimestampToCurrentTime() {
+    oppiaClock.setFakeTimeMode(FakeOppiaClock.FakeTimeMode.MODE_FIXED_FAKE_TIME)
+    explorationActiveTimeController.setExplorationSessionStarted()
+    val currentTime = oppiaClock.getCurrentTimeMs()
+    assertThat(explorationActiveTimeController.startExplorationTimestampMs).isEqualTo(currentTime)
+  }
+
+  @Test
+  fun testRecordAggregateTopicLearningTime_returnsSuccess() {
+    val recordAggregateTimeProvider =
+      explorationActiveTimeController.recordAggregateTopicLearningTime(
+        profileId, TEST_TOPIC_ID_0, SESSION_LENGTH
+      )
+    monitorFactory.waitForNextSuccessfulResult(recordAggregateTimeProvider)
+  }
+
+  @Test
+  fun testSaveSessionLength_previousAggregateIsStale_overwritesOldAggregateWithCurrentLength() {
```

Comment:
This test always fails and I am unable to figure out why.
I have tried to debug using breakpoints, and the the code always hits the wrong if/else branch -- i.e the one that evaluates to timestamp not stale, and I cannot tell whether it is an issue with test set up or API. The API looks correct to me.

Reply:
As discussed & found during the meeting, it seems the failure is due to the PersistentCacheStore not actually writing the expected contents ahead of expecting the value to be available. I also suggest following a pattern more like AppStartupStateControllerTest wherein we only use the ExplorationActiveTimeController's API for setting & verifying state (since we can use two instances of the test suite's TestApplicationComponent to simulate existing state from a "previous app run").

Reply:
I was able to refactor all the tests to use the public apis by simulating the previous run suggested. Thanks.